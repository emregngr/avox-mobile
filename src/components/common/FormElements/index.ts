export * from './CheckboxField'
export * from './TextInputField'
