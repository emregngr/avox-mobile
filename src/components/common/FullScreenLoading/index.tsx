import React from 'react'
import { ActivityIndicator, View } from 'react-native'

export const FullScreenLoading = () => (
  <View
    className="flex-1 items-center justify-center bg-background-primary"
    testID="full-screen-loading"
  >
    <ActivityIndicator className="color-text-100" size="large" testID="loading-indicator" />
  </View>
)
