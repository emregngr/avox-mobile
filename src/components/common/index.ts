export * from './FaqItem'
export * from './FormElements'
export * from './FullScreenLoading'
export * from './Header'
export * from './OnboardingItem'
export * from './ProfileItem'
export * from './RenderDetailTabBar'
export * from './RenderTabBar'
export * from './SafeLayout'
export * from './SearchInput'
export * from './ThemedButton'
export * from './ThemedButtonText'
export * from './ThemedGradientButton'
export * from './ThemedText'
export * from './ToastConfig'
