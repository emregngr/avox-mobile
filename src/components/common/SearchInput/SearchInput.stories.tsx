import type { Meta, StoryObj } from '@storybook/react-native'
import React, { useEffect, useState } from 'react'
import { Keyboard, ScrollView, TouchableOpacity, View } from 'react-native'

import { SearchInput } from '@/components/common/SearchInput'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

const SearchInputDemo = ({
  placeholder = 'Search airports, airlines...',
  showState = true,
}: {
  placeholder?: string
  showState?: boolean
}) => {
  const { selectedTheme } = useThemeStore()
  const colors = themeColors?.[selectedTheme]

  const [searchValue, setSearchValue] = useState<string>('')
  const [searchHistory, setSearchHistory] = useState<string[]>([])

  const handleSearch = (text: string) => {
    setSearchValue(text)
    if (text.trim() && !searchHistory.includes(text.trim())) {
      setSearchHistory(prev => [text.trim(), ...prev.slice(0, 4)])
    }
  }

  const handleHistorySelect = (item: string) => {
    setSearchValue(item)
  }

  return (
    <View className="flex-1 bg-background-primary pt-5">
      <SearchInput value={searchValue} onChangeText={handleSearch} placeholder={placeholder} />

      {showState ? (
        <View className="flex-1 px-4 pt-5">
          <View className="bg-background-secondary rounded-xl p-4 mb-5">
            <ThemedText color="text-100" type="body1" className="mb-2">
              Current Value:
            </ThemedText>
            <View className="bg-background-secondary p-3 rounded-lg mb-4">
              <ThemedText color={searchValue ? 'text-100' : 'text-70'} type="body2">
                {searchValue || 'No input'}
              </ThemedText>
            </View>

            <ThemedText color="text-100" type="body1" className="mb-2">
              Character Count: {searchValue.length}
            </ThemedText>

            <View
              className="p-2 rounded-md items-center"
              style={{
                backgroundColor:
                  searchValue.length > 0
                    ? colors?.background?.secondary
                    : colors?.background?.tertiary,
              }}
            >
              <ThemedText color="text-100" type="h3">
                {searchValue.length > 0 ? 'Has Content' : 'Empty'}
              </ThemedText>
            </View>
          </View>

          {searchHistory.length > 0 ? (
            <View className="bg-background-secondary rounded-xl p-4">
              <ThemedText color="text-100" type="body1" className="mb-3">
                Recent Searches:
              </ThemedText>
              {searchHistory.map((item, index) => (
                <TouchableOpacity
                  key={index}
                  activeOpacity={0.7}
                  className="bg-background-secondary p-3 rounded-lg mb-2"
                  onPress={() => handleHistorySelect(item)}
                >
                  <ThemedText color="text-70" type="body2">
                    {item}
                  </ThemedText>
                </TouchableOpacity>
              ))}
            </View>
          ) : null}
        </View>
      ) : null}
    </View>
  )
}

const meta: Meta<typeof SearchInput> = {
  title: 'Components/SearchInput',
  component: SearchInput,
  parameters: {
    docs: {
      description: {
        component:
          'Animated search input component with focus states, clear functionality, and cancel button. Features responsive design and theme integration.',
      },
    },
  },
  decorators: [
    Story => (
      <View className="flex-1 bg-background-primary">
        <Story />
      </View>
    ),
  ],
  argTypes: {
    placeholder: {
      control: { type: 'text' },
      description: 'Placeholder text for the search input',
    },
    value: {
      control: { type: 'text' },
      description: 'Current input value',
    },
    className: {
      control: { type: 'text' },
      description: 'Additional CSS classes',
    },
  },
}

export default meta

type Story = StoryObj<typeof SearchInputDemo>

export const Default: Story = {
  render: () => <SearchInputDemo />,
  parameters: {
    docs: {
      description: {
        story: 'Default search input with real-time state display and search history.',
      },
    },
  },
}

export const PlaceholderVariants: Story = {
  render: () => {
    const [airportSearchValue, setAirportSearchValue] = useState<string>('')
    const [airlineSearchValue, setAirlineSearchValue] = useState<string>('')
    const [generalSearchValue, setGeneralSearchValue] = useState<string>('')
    const [locationSearchValue, setLocationSearchValue] = useState<string>('')

    return (
      <ScrollView
        contentContainerClassName="flex-1 pt-5 pb-10 gap-4"
        showsVerticalScrollIndicator={false}
      >
        <View>
          <ThemedText color="text-100" type="h4" className="mb-2.5 ml-4">
            Airport Search
          </ThemedText>
          <SearchInput
            value={airportSearchValue}
            onChangeText={setAirportSearchValue}
            placeholder="Search airports..."
          />
        </View>

        <View>
          <ThemedText color="text-100" type="h4" className="mb-2.5 ml-4">
            Airline Search
          </ThemedText>
          <SearchInput
            value={airlineSearchValue}
            onChangeText={setAirlineSearchValue}
            placeholder="Search airlines..."
          />
        </View>

        <View>
          <ThemedText color="text-100" type="h4" className="mb-2.5 ml-4">
            General Search
          </ThemedText>
          <SearchInput
            value={generalSearchValue}
            onChangeText={setGeneralSearchValue}
            placeholder="What are you looking for?"
          />
        </View>

        <View>
          <ThemedText color="text-100" type="h4" className="mb-2.5 ml-4">
            Location Search
          </ThemedText>
          <SearchInput
            value={locationSearchValue}
            onChangeText={setLocationSearchValue}
            placeholder="Enter city, airport code..."
          />
        </View>
      </ScrollView>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Different placeholder text variations for various search contexts.',
      },
    },
  },
}

export const FocusStates: Story = {
  render: () => {
    const [focusedInput, setFocusedInput] = useState<number | null>(null)
    const [values, setValues] = useState<string[]>(['', '', ''])

    const createHandleChange = (index: number) => (text: string) => {
      const newValues = [...values]
      newValues[index] = text
      setValues(newValues)
    }

    return (
      <ScrollView
        contentContainerClassName="flex-1 pt-5 pb-10 gap-4"
        showsVerticalScrollIndicator={false}
      >
        <View>
          <ThemedText color="text-100" type="body1" className="mb-2.5 ml-4">
            Input 1 {focusedInput === 0 ? '(Focused)' : '(Unfocused)'}
          </ThemedText>
          <SearchInput
            value={values?.[0] as string}
            onChangeText={createHandleChange(0)}
            placeholder="Focus to see changes..."
          />
        </View>

        <View>
          <ThemedText color="text-100" type="body1" className="mb-2.5 ml-4">
            Input 2 {focusedInput === 1 ? '(Focused)' : '(Unfocused)'}
          </ThemedText>
          <SearchInput
            value={values?.[1] as string}
            onChangeText={createHandleChange(1)}
            placeholder="Type something here..."
          />
        </View>

        <View>
          <ThemedText color="text-100" type="body1" className="mb-2.5 ml-4">
            Input 3 {focusedInput === 2 ? '(Focused)' : '(Unfocused)'}
          </ThemedText>
          <SearchInput
            value={values?.[2] as string}
            onChangeText={createHandleChange(2)}
            placeholder="Another search input..."
          />
        </View>

        <TouchableOpacity
          activeOpacity={0.7}
          className="bg-success p-4 rounded-xl mx-4 items-center"
          onPress={() => Keyboard.dismiss()}
        >
          <ThemedText color="text-100" type="body1">
            Dismiss Keyboard
          </ThemedText>
        </TouchableOpacity>

        <View className="p-4 bg-background-secondary rounded-xl mx-4">
          <ThemedText color="text-100" type="body1" className="mb-3">
            Focus State Features:
          </ThemedText>
          <ThemedText color="text-70" type="body2">
            • Input expands when focused{'\n'}• Cancel button appears on focus{'\n'}• Clear button
            shows with content{'\n'}• Smooth animations between states{'\n'}• Auto keyboard
            appearance matching theme
          </ThemedText>
        </View>
      </ScrollView>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Focus state behavior and animations with multiple inputs.',
      },
    },
  },
}

export const InteractiveFeatures: Story = {
  render: () => {
    const [searchValue, setSearchValue] = useState<string>('Sample search text')
    const [actionLog, setActionLog] = useState<string[]>([])

    const logAction = (action: string) => {
      const timestamp = new Date().toLocaleTimeString()
      setActionLog(prev => [`${timestamp}: ${action}`, ...prev.slice(0, 9)])
    }

    const handleSearch = (text: string) => {
      setSearchValue(text)
      logAction(`Text changed to: "${text}"`)
    }

    const clearLog = () => {
      setActionLog([])
      logAction('Action log cleared')
    }

    return (
      <View className="flex-1 pt-5">
        <SearchInput
          value={searchValue}
          onChangeText={handleSearch}
          placeholder="Interactive search demo..."
        />

        <ScrollView className="flex-1 px-4 pt-5" showsVerticalScrollIndicator={false}>
          <View className="bg-background-secondary rounded-xl p-4 mb-5">
            <ThemedText color="text-100" type="h4" className="mb-4">
              Interactive Features
            </ThemedText>

            <View className="mb-4">
              <ThemedText color="success" type="body2" className="mb-2">
                🔍 Search Icon: Always visible
              </ThemedText>
              <ThemedText color="text-70" type="h3">
                Magnify icon using themed colors
              </ThemedText>
            </View>

            <View className="mb-4">
              <ThemedText color="warning" type="body2" className="mb-2">
                ❌ Clear Button: Shows when focused + has content
              </ThemedText>
              <ThemedText color="text-70" type="h3">
                Custom close icon with primary/secondary colors
              </ThemedText>
            </View>

            <View className="mb-4">
              <ThemedText color="error" type="body2" className="mb-2">
                🚫 Cancel Button: Shows when focused
              </ThemedText>
              <ThemedText color="text-70" type="h3">
                Clears text and dismisses keyboard
              </ThemedText>
            </View>

            <View>
              <ThemedText color="primary-100" type="body2" className="mb-2">
                📱 Keyboard: Theme-aware appearance
              </ThemedText>
              <ThemedText color="text-70" type="h3">
                Matches selected theme (light/dark)
              </ThemedText>
            </View>
          </View>

          <View className="bg-background-secondary rounded-xl p-4 mb-5">
            <View className="flex-row justify-between items-center mb-3">
              <ThemedText color="text-100" type="body1">
                Action Log ({actionLog.length})
              </ThemedText>
              <TouchableOpacity
                activeOpacity={0.7}
                className="bg-error px-3 py-1.5 rounded-md"
                onPress={clearLog}
              >
                <ThemedText color="text-100" type="h3">
                  Clear
                </ThemedText>
              </TouchableOpacity>
            </View>

            <View>
              <ScrollView showsVerticalScrollIndicator={false}>
                {actionLog.length === 0 ? (
                  <ThemedText color="text-70" type="body2">
                    No actions logged yet
                  </ThemedText>
                ) : (
                  actionLog.map((log, index) => (
                    <View key={index} className="bg-background-primary p-2 rounded mb-1">
                      <ThemedText color="text-70" type="h3">
                        {log}
                      </ThemedText>
                    </View>
                  ))
                )}
              </ScrollView>
            </View>
          </View>
        </ScrollView>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive features demonstration with action logging.',
      },
    },
  },
}

export const ResponsiveLayout: Story = {
  render: () => {
    const [searchValue, setSearchValue] = useState<string>('')

    return (
      <View className="flex-1 pt-5">
        <SearchInput
          value={searchValue}
          onChangeText={setSearchValue}
          placeholder="Responsive search input"
        />

        <View className="p-4 bg-background-secondary rounded-xl mx-4 mt-5">
          <ThemedText color="text-100" type="h4" className="mb-4">
            Responsive Layout
          </ThemedText>

          <ThemedText color="text-70" type="body2" className="mb-4">
            The search input automatically adapts to different screen sizes:
          </ThemedText>

          <View className="mb-3">
            <ThemedText color="info" type="body2" className="mb-1">
              📐 Width Calculation:
            </ThemedText>
            <ThemedText color="text-70" type="h3">
              Device width - 32px (16px padding on each side)
            </ThemedText>
          </View>

          <View className="mb-3">
            <ThemedText color="success" type="body2" className="mb-1">
              📱 Self-Centering:
            </ThemedText>
            <ThemedText color="text-70" type="h3">
              Automatically centers within its container
            </ThemedText>
          </View>

          <View className="mb-3">
            <ThemedText color="warning" type="body2" className="mb-1">
              🔄 Dynamic Expansion:
            </ThemedText>
            <ThemedText color="text-70" type="h3">
              Input field expands when focused, cancel button appears
            </ThemedText>
          </View>

          <View>
            <ThemedText color="error" type="body2" className="mb-1">
              📏 Consistent Height:
            </ThemedText>
            <ThemedText color="text-70" type="h3">
              Maintains consistent height across all screen sizes
            </ThemedText>
          </View>
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Responsive layout behavior across different screen sizes.',
      },
    },
  },
}

export const AccessibilityTesting: Story = {
  render: () => {
    const [searchValue, setSearchValue] = useState<string>('')
    const [testResults, setTestResults] = useState<string[]>([])

    const runAccessibilityTests = () => {
      const tests = [
        'Search input has testID: search-input ✅',
        'Clear button has testID: search-clear-button ✅',
        'Cancel button has testID: search-cancel-button ✅',
        'Font scaling disabled for consistency ✅',
        'Keyboard appearance matches theme ✅',
        'Proper hit slop for touch targets ✅',
        'Return key type set to "search" ✅',
        'Auto-correct and spell check disabled ✅',
      ]

      setTestResults(tests)
    }

    return (
      <View className="flex-1 pt-5">
        <SearchInput
          value={searchValue}
          onChangeText={setSearchValue}
          placeholder="Accessibility test input..."
        />

        <ScrollView className="flex-1 px-4 pt-5" showsVerticalScrollIndicator={false}>
          <View className="bg-background-secondary rounded-xl p-4 mb-5">
            <ThemedText color="text-100" type="h4" className="mb-4">
              Test Identifiers
            </ThemedText>

            <View className="bg-background-primary p-3 rounded-lg mb-4">
              <ThemedText color="success" type="body2" className="mb-2">
                Available Test IDs:
              </ThemedText>
              <ThemedText color="text-70" type="h3">
                search-input{'\n'}
                search-clear-button{'\n'}
                search-cancel-button
              </ThemedText>
            </View>

            <TouchableOpacity
              activeOpacity={0.7}
              className="bg-success p-3 rounded-lg items-center"
              onPress={runAccessibilityTests}
            >
              <ThemedText color="text-100" type="body2">
                Run Accessibility Tests
              </ThemedText>
            </TouchableOpacity>
          </View>

          {testResults.length > 0 ? (
            <View className="bg-background-secondary rounded-xl p-4 mb-5">
              <ThemedText color="text-100" type="h4" className="mb-4">
                Test Results
              </ThemedText>

              {testResults.map((result, index) => (
                <View key={index} className="bg-background-primary p-2.5 rounded-md mb-1.5">
                  <ThemedText color="success" type="h3">
                    {result}
                  </ThemedText>
                </View>
              ))}
            </View>
          ) : null}

          <View className="bg-background-secondary rounded-xl p-4">
            <ThemedText color="text-100" type="body1" className="mb-3">
              Accessibility Features:
            </ThemedText>
            <ThemedText color="text-70" type="body2">
              • Disabled font scaling for UI consistency{'\n'}• Proper hit slop for touch targets
              {'\n'}• Theme-aware keyboard appearance{'\n'}• Semantic return key type (search){'\n'}
              • Disabled auto-correct for search context{'\n'}• Transparent underline for clean
              appearance{'\n'}• Test IDs for automated testing
            </ThemedText>
          </View>
        </ScrollView>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Accessibility features and testing capabilities.',
      },
    },
  },
}

export const PerformanceAnimation: Story = {
  render: () => {
    const [searchValue, setSearchValue] = useState<string>('')
    const [renderCount, setRenderCount] = useState<number>(0)
    const [animationState, setAnimationState] = useState<'idle' | 'typing'>('idle')

    useEffect(() => {
      setRenderCount(prev => prev + 1)
    }, [searchValue])

    const handleChange = (text: string) => {
      setSearchValue(text)
      setAnimationState('typing')

      setTimeout(() => setAnimationState('idle'), 300)
    }

    const stressTest = () => {
      let count = 0
      const interval = setInterval(() => {
        setSearchValue(`Test ${count}`)
        count++
        if (count > 20) {
          clearInterval(interval)
          setSearchValue('')
        }
      }, 100)
    }

    return (
      <View className="flex-1 pt-5">
        <SearchInput
          value={searchValue}
          onChangeText={handleChange}
          placeholder="Performance test input..."
        />

        <ScrollView className="flex-1 px-4 pt-5" showsVerticalScrollIndicator={false}>
          <View className="bg-background-secondary rounded-xl p-4 mb-5">
            <ThemedText color="text-100" type="h4" className="mb-4">
              Performance Metrics
            </ThemedText>

            <View className="flex-row justify-between mb-3">
              <ThemedText color="text-90" type="body2">
                Render Count:
              </ThemedText>
              <ThemedText color="primary-100" type="body2">
                {renderCount}
              </ThemedText>
            </View>

            <View className="flex-row justify-between mb-3">
              <ThemedText color="text-90" type="body2">
                Animation State:
              </ThemedText>
              <ThemedText color={animationState === 'typing' ? 'warning' : 'success'} type="body2">
                {animationState}
              </ThemedText>
            </View>

            <View className="flex-row justify-between mb-4">
              <ThemedText color="text-90" type="body2">
                Input Length:
              </ThemedText>
              <ThemedText color="text-100" type="body2">
                {searchValue.length} chars
              </ThemedText>
            </View>

            <TouchableOpacity
              activeOpacity={0.7}
              className="bg-background-secondary p-3 rounded-lg items-center"
              onPress={stressTest}
            >
              <ThemedText color="text-100" type="body2">
                Run Stress Test (20 rapid changes)
              </ThemedText>
            </TouchableOpacity>
          </View>

          <View className="bg-background-secondary rounded-xl p-4">
            <ThemedText color="text-100" type="body1" className="mb-3">
              Optimization Features:
            </ThemedText>
            <ThemedText color="text-70" type="body2">
              • useMemo for theme colors{'\n'}• useRef for TextInput reference{'\n'}• Optimized
              re-renders{'\n'}• CSS transition animations{'\n'}• Conditional rendering for buttons
              {'\n'}• Efficient state management{'\n'}• Native driver animations where possible
            </ThemedText>
          </View>
        </ScrollView>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Performance monitoring and animation testing capabilities.',
      },
    },
  },
}
