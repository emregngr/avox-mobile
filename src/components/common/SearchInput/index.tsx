import { MaterialCommunityIcons } from '@expo/vector-icons'
import React, { useMemo, useRef, useState } from 'react'
import { TextInput, TouchableOpacity, View } from 'react-native'

import Close from '@/assets/icons/close'
import { ThemedButtonText } from '@/components/common/ThemedButtonText'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { cn } from '@/utils/common/cn'
import { responsive } from '@/utils/common/responsive'

interface SearchInputProps {
  className?: string
  onChangeText: (text: string) => void
  placeholder: string
  value: string
}

export const SearchInput = ({ className, onChangeText, placeholder, value }: SearchInputProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const textInputRef = useRef<TextInput>(null)

  const [isFocused, setIsFocused] = useState<boolean>(false)

  const handleClear = () => {
    onChangeText('')
  }

  const handleCancel = () => {
    onChangeText('')
    setIsFocused(false)
    textInputRef.current?.blur()
  }

  return (
    <View
      style={{
        width: responsive.deviceWidth - 32,
      }}
      className={cn('flex-row items-center self-center', className)}
    >
      <View
        className={cn(
          'flex-row items-center px-4 rounded-xl overflow-hidden bg-background-tertiary transition-all duration-300',
          isFocused ? 'flex-1' : '',
        )}
      >
        <MaterialCommunityIcons color={colors?.onPrimary100} name="magnify" size={20} />
        <TextInput
          onBlur={() => {
            setIsFocused(false)
          }}
          allowFontScaling={false}
          autoCorrect={false}
          className="flex-1 py-3 ml-3 text-text-100 placeholder:text-text-50 text-[16px] font-inter-medium"
          keyboardAppearance={selectedTheme}
          maxFontSizeMultiplier={1.0}
          onChangeText={onChangeText}
          onFocus={() => setIsFocused(true)}
          placeholder={placeholder}
          ref={textInputRef}
          returnKeyType="search"
          spellCheck={false}
          testID="search-input"
          textAlignVertical="center"
          underlineColorAndroid="transparent"
          value={value}
          enablesReturnKeyAutomatically
        />
        {isFocused && value?.length > 0 ? (
          <TouchableOpacity
            activeOpacity={0.7}
            hitSlop={10}
            onPress={handleClear}
            testID="search-clear-button"
          >
            <Close
              height={20}
              primaryColor={colors?.background?.primary}
              secondaryColor={colors?.onPrimary100}
              width={20}
            />
          </TouchableOpacity>
        ) : null}
      </View>
      {isFocused ? (
        <ThemedButtonText
          containerStyle="ml-2"
          hitSlop={10}
          label={getLocale('cancel')}
          onPress={handleCancel}
          testID="search-cancel-button"
          textColor="text-100"
          type="body1"
        />
      ) : null}
    </View>
  )
}
