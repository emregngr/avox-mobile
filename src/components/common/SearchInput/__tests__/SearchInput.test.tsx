import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'

import { SearchInput } from '@/components/common/SearchInput'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'

jest.mock('@/locales/i18next')

const mockedGetLocale = getLocale as jest.MockedFunction<typeof getLocale>

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

jest.mock('@/assets/icons/close', () => 'Close')

jest.mock('@/components/common/ThemedButtonText', () => {
  const { TouchableOpacity } = require('react-native')

  return {
    ThemedButtonText: (props: any) => <TouchableOpacity {...props} />,
  }
})

jest.mock('@/utils/common/cn', () => ({
  cn: jest.fn((...classes) => classes.filter(Boolean).join(' ')),
}))

jest.mock('@/utils/common/responsive', () => ({
  responsive: {
    deviceWidth: 375,
  },
}))

const mockedDefaultProps = {
  onChangeText: jest.fn(),
  placeholder: 'Search...',
  value: '',
}

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({
    selectedTheme: 'light',
  })

  mockedGetLocale.mockImplementation((key: string) => {
    const translations: Record<string, string> = {
      cancel: 'Cancel',
    }
    return translations[key] || key
  })
})

describe('SearchInput Component', () => {
  describe('Rendering', () => {
    it('renders correctly with default props', () => {
      const { getByPlaceholderText } = render(<SearchInput {...mockedDefaultProps} />)

      expect(getByPlaceholderText('Search...')).toBeTruthy()
    })

    it('renders with custom className', () => {
      render(<SearchInput {...mockedDefaultProps} className="custom-class" />)

      expect(require('@/utils/common/cn').cn).toHaveBeenCalledWith(
        'flex-row items-center self-center',
        'custom-class',
      )
    })

    it('displays the correct placeholder text', () => {
      const { getByPlaceholderText } = render(
        <SearchInput {...mockedDefaultProps} placeholder="Custom placeholder" />,
      )

      expect(getByPlaceholderText('Custom placeholder')).toBeTruthy()
    })

    it('displays the current value', () => {
      const { getByDisplayValue } = render(
        <SearchInput {...mockedDefaultProps} value="test value" />,
      )

      const textInput = getByDisplayValue('test value')
      expect(textInput).toBeTruthy()
    })
  })

  describe('Focus behavior', () => {
    it('shows cancel button when focused', () => {
      const { getByPlaceholderText, getByTestId } = render(<SearchInput {...mockedDefaultProps} />)

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      expect(getByTestId('search-cancel-button')).toBeTruthy()
    })

    it('hides cancel button when blurred', () => {
      const { getByPlaceholderText, queryByTestId } = render(
        <SearchInput {...mockedDefaultProps} />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')
      fireEvent(textInput, 'blur')

      expect(queryByTestId('ThemedButtonText')).toBeFalsy()
    })

    it('applies focused styles when input is focused', () => {
      const { getByPlaceholderText } = render(<SearchInput {...mockedDefaultProps} />)

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      expect(require('@/utils/common/cn').cn).toHaveBeenCalledWith(
        'flex-row items-center px-4 rounded-xl overflow-hidden bg-background-tertiary transition-all duration-300',
        'flex-1',
      )
    })
  })

  describe('Clear functionality', () => {
    it('shows clear button when focused and has value', () => {
      const { getByPlaceholderText, getByTestId } = render(
        <SearchInput {...mockedDefaultProps} value="test" />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      expect(getByTestId('search-clear-button')).toBeTruthy()
    })

    it('hides clear button when not focused', () => {
      const { queryByTestId } = render(<SearchInput {...mockedDefaultProps} value="test" />)

      expect(queryByTestId('search-clear-button')).toBeFalsy()
    })

    it('hides clear button when focused but no value', () => {
      const { getByPlaceholderText, queryByTestId } = render(
        <SearchInput {...mockedDefaultProps} value="" />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      expect(queryByTestId('search-clear-button')).toBeFalsy()
    })

    it('calls onChangeText with empty string when clear button is pressed', () => {
      const mockedOnChangeText = jest.fn()
      const { getByPlaceholderText, getByTestId } = render(
        <SearchInput {...mockedDefaultProps} onChangeText={mockedOnChangeText} value="test" />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      const clearButton = getByTestId('search-clear-button')
      fireEvent.press(clearButton)

      expect(mockedOnChangeText).toHaveBeenCalledWith('')
    })
  })

  describe('Cancel functionality', () => {
    it('calls onChangeText with empty string when cancel is pressed', () => {
      const mockedOnChangeText = jest.fn()
      const { getByPlaceholderText, getByTestId } = render(
        <SearchInput {...mockedDefaultProps} onChangeText={mockedOnChangeText} />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      const cancelButton = getByTestId('search-cancel-button')
      fireEvent.press(cancelButton)

      expect(mockedOnChangeText).toHaveBeenCalledWith('')
    })

    it('blurs the input when cancel is pressed', () => {
      const { getByPlaceholderText, getByTestId, queryByTestId } = render(
        <SearchInput {...mockedDefaultProps} />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      const cancelButton = getByTestId('search-cancel-button')
      fireEvent.press(cancelButton)

      expect(queryByTestId('ThemedButtonText')).toBeFalsy()
    })
  })

  describe('Text input behavior', () => {
    it('calls onChangeText when text is entered', () => {
      const mockedOnChangeText = jest.fn()
      const { getByPlaceholderText } = render(
        <SearchInput {...mockedDefaultProps} onChangeText={mockedOnChangeText} />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent.changeText(textInput, 'new text')

      expect(mockedOnChangeText).toHaveBeenCalledWith('new text')
    })

    it('has correct text input properties', () => {
      const { getByPlaceholderText } = render(<SearchInput {...mockedDefaultProps} />)

      const textInput = getByPlaceholderText('Search...')

      expect(textInput.props.allowFontScaling).toBe(false)
      expect(textInput.props.autoCorrect).toBe(false)
      expect(textInput.props.returnKeyType).toBe('search')
      expect(textInput.props.spellCheck).toBe(false)
      expect(textInput.props.enablesReturnKeyAutomatically).toBe(true)
    })
  })

  describe('Theme integration', () => {
    it('passes keyboard appearance based on selected theme', () => {
      const { getByPlaceholderText } = render(<SearchInput {...mockedDefaultProps} />)

      const textInput = getByPlaceholderText('Search...')
      expect(textInput.props.keyboardAppearance).toBe('light')
    })
  })

  describe('Accessibility', () => {
    it('has proper hit slop for touchable elements', () => {
      const { getByPlaceholderText, getByTestId } = render(
        <SearchInput {...mockedDefaultProps} value="test" />,
      )

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      const clearButton = getByTestId('search-clear-button')

      const parentElement = clearButton.parent?.parent || clearButton.parent
      expect(parentElement?.props?.hitSlop).toBe(10)
    })

    it('cancel button has proper hit slop', () => {
      const { getByPlaceholderText, getByTestId } = render(<SearchInput {...mockedDefaultProps} />)

      const textInput = getByPlaceholderText('Search...')
      fireEvent(textInput, 'focus')

      const cancelButton = getByTestId('search-cancel-button')
      expect(cancelButton.props.hitSlop).toBe(10)
    })
  })

  describe('Layout', () => {
    it('sets correct width based on device width', () => {
      const { root } = render(<SearchInput {...mockedDefaultProps} />)

      const findElementWithWidth = (element: any): any => {
        if (element.props?.style?.width === 343) {
          return element
        }
        if (element.children) {
          for (const child of element.children) {
            const found = findElementWithWidth(child)
            if (found) return found
          }
        }
        return null
      }

      const elementWithWidth = findElementWithWidth(root)
      expect(elementWithWidth).toBeTruthy()
      expect(elementWithWidth.props.style.width).toBe(343)
    })
  })
})

describe('SearchInput Component Integration', () => {
  it('handles complete user interaction flow', () => {
    const mockedOnChangeText = jest.fn()
    const { rerender, getByPlaceholderText, getByTestId } = render(
      <SearchInput {...mockedDefaultProps} onChangeText={mockedOnChangeText} />,
    )

    const textInput = getByPlaceholderText('Search...')

    fireEvent(textInput, 'focus')
    expect(getByTestId('search-cancel-button')).toBeTruthy()

    fireEvent.changeText(textInput, 'search query')
    expect(mockedOnChangeText).toHaveBeenCalledWith('search query')

    rerender(
      <SearchInput
        {...mockedDefaultProps}
        onChangeText={mockedOnChangeText}
        value="search query"
      />,
    )

    fireEvent(textInput, 'focus')

    expect(getByTestId('search-clear-button')).toBeTruthy()

    fireEvent.press(getByTestId('search-clear-button'))
    expect(mockedOnChangeText).toHaveBeenCalledWith('')

    fireEvent.press(getByTestId('search-cancel-button'))
    expect(mockedOnChangeText).toHaveBeenCalledWith('')
  })
})

describe('SearchInput Component Snapshot', () => {
  it('should render the SearchInput Component successfully', () => {
    const { toJSON } = render(<SearchInput {...mockedDefaultProps} />)

    expect(toJSON()).toMatchSnapshot()
  })
})
