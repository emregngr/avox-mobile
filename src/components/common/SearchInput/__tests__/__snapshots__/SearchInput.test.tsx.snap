// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SearchInput Component Snapshot should render the SearchInput Component successfully 1`] = `
<View
  style={
    {
      "width": 343,
    }
  }
>
  <View>
    <View
      color="#000000"
      name="magnify"
      size={20}
      testID="mocked-material-community-icon"
    />
    <TextInput
      allowFontScaling={false}
      autoCorrect={false}
      className="flex-1 py-3 ml-3 text-text-100 placeholder:text-text-50 text-[16px] font-inter-medium"
      enablesReturnKeyAutomatically={true}
      keyboardAppearance="light"
      maxFontSizeMultiplier={1}
      onBlur={[Function]}
      onChangeText={[MockFunction]}
      onFocus={[Function]}
      placeholder="Search..."
      returnKeyType="search"
      spellCheck={false}
      testID="search-input"
      textAlignVertical="center"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
